haskell.nix revamp

angerman opened the pull request:
Ok. This is basically a full haskell.nix restart from the grounds up. We'll need to add back the control logic as well. But I'd like this to be a process with @erikd, so he is part of the discovery what is ci related, what is operations related, ... and can have a word on where that logic should reside.

Task list:
- [x] default.nix (haskell.nix setup to build the project form the `cabal.project` as ground truth)
- [x] ~ci.nix (to build the project using hercules-ci)~ We'll use `default.nix` only for now.
- [x] on `sha256` why we need them and how to update them.
- [x] bring back docker logic
- [x] bring back deployment/ops logic.

angerman committed:
drop existing nix code

angerman committed:
Adds default.nix

angerman committed:
add hashes and extra deps to cabal.project

add io-sim, ouroboros-network-testing, cardano-shell. We need those
for haskell.nix as it tries to compute a full plan (with all tests).

erikd commented:
Would like to see this lot squashed down to two commits, one with the changes to `cabal.project` /`stack.yaml` and the other with all the Nix changes.

angerman committed:
Add niv (nix source reference manager) back

angerman committed:
Add nixOS services (explorer-frontend, exporter and graphql) back.

These are supposed to be imported by a nixOS machine to make these
services available.

angerman committed:
Add shell and scripts back

angerman committed:
Add postgresql service to test-db test case

angerman commented:
@erikd alirght, I think what's missing is the docker stuff, I'll try to finish that tomorrow, completely missed that piece.

Please ask all the questions about nix you have about this. It cuts both ways. The pure CI logic essenitally ends before the `Add niv` commit. That one and the following is nix code related to operational concerns (providing a service description, ...)

angerman committed:
Bring back `niv-shell` (ops)

angerman committed:
Add ci.nix (hercules-ci control file)

This will add
- x86_64-linux
- x86_64-macos
- x86_64-mingw (on linux)
- docker inputs

angerman committed:
kill postgres at the end (fixes darwin)

angerman committed:
[fix docker] drop haskellPackages

angerman committed:
fix windows?

angerman commented on update-cabal-project.py:
```diff
@@ -0,0 +1,58 @@
+    dict["subdir"] = ''
+  if not (dict['loc'], dict['tag']) in globalCache:
+    prefetchJSON = subprocess.run(
+      ["nix-prefetch-git", "--quiet", dict['loc'], dict['tag']],
```
This one should have an option to go through ssh

angerman committed:
Allow to run `nix-prefetch-git` via ssh.

angerman committed:
bump haskell.nix

angerman committed:
fix default.nix; use buildPackages
